File: OpenHome/Net/Bindings/Cs/ControlPoint/Proxies/CpAvOpenhomeOrgMessageCenter1.cs
using System;
using System.Runtime.InteropServices;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using OpenHome.Net.Core;
using OpenHome.Net.ControlPoint;

namespace OpenHome.Net.ControlPoint.Proxies
{
    public interface ICpProxyAvOpenhomeOrgMessageCenter1 : ICpProxy, IDisposable
    {
        void SyncGetMessage(out String aMessage, out uint aMessageID);
        void BeginGetMessage(CpProxy.CallbackAsyncComplete aCallback);
        void EndGetMessage(IntPtr aAsyncHandle, out String aMessage, out uint aMessageID);
        void SetPropertyMessageChanged(System.Action aMessageChanged);
        String PropertyMessage();
        void SetPropertyMessageIDChanged(System.Action aMessageIDChanged);
        uint PropertyMessageID();
    }

    internal class SyncGetMessageAvOpenhomeOrgMessageCenter1 : SyncProxyAction
    {
        private CpProxyAvOpenhomeOrgMessageCenter1 iService;
        private String iMessage;
        private uint iMessageID;

        public SyncGetMessageAvOpenhomeOrgMessageCenter1(CpProxyAvOpenhomeOrgMessageCenter1 aProxy)
        {
            iService = aProxy;
        }
        public String Message()
        {
            return iMessage;
        }
        public uint MessageID()
        {
            return iMessageID;
        }
        protected override void CompleteRequest(IntPtr aAsyncHandle)
        {
            iService.EndGetMessage(aAsyncHandle, out iMessage, out iMessageID);
        }
    };

    /// <summary>
    /// Proxy for the av.openhome.org:MessageCenter:1 UPnP service
    /// </summary>
    public class CpProxyAvOpenhomeOrgMessageCenter1 : CpProxy, IDisposable, ICpProxyAvOpenhomeOrgMessageCenter1
    {
        private OpenHome.Net.Core.Action iActionGetMessage;
        private PropertyString iMessage;
        private PropertyUint iMessageID;
        private System.Action iMessageChanged;
        private System.Action iMessageIDChanged;
        private Mutex iPropertyLock;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <remarks>Use CpProxy::[Un]Subscribe() to enable/disable querying of state variable and reporting of their changes.</remarks>
        /// <param name="aDevice">The device to use</param>
        public CpProxyAvOpenhomeOrgMessageCenter1(CpDevice aDevice)
            : base("av-openhome-org", "MessageCenter", 1, aDevice)
        {
            OpenHome.Net.Core.Parameter param;
            List<String> allowedValues = new List<String>();

            iActionGetMessage = new OpenHome.Net.Core.Action("GetMessage");
            param = new ParameterString("Message", allowedValues);
            iActionGetMessage.AddOutputParameter(param);
            param = new ParameterUint("MessageID");
            iActionGetMessage.AddOutputParameter(param);

            iMessage = new PropertyString("Message", MessagePropertyChanged);
            AddProperty(iMessage);
            iMessageID = new PropertyUint("MessageID", MessageIDPropertyChanged);
            AddProperty(iMessageID);
            
            iPropertyLock = new Mutex();
        }

        /// <summary>
        /// Invoke the action synchronously
        /// </summary>
        /// <remarks>Blocks until the action has been processed
        /// on the device and sets any output arguments</remarks>
        /// <param name="aMessage"></param>
        /// <param name="aMessageID"></param>
        public void SyncGetMessage(out String aMessage, out uint aMessageID)
        {
            SyncGetMessageAvOpenhomeOrgMessageCenter1 sync = new SyncGetMessageAvOpenhomeOrgMessageCenter1(this);
            BeginGetMessage(sync.AsyncComplete());
            sync.Wait();
            sync.ReportError();
            aMessage = sync.Message();
            aMessageID = sync.MessageID();
        }

        /// <summary>
        /// Invoke the action asynchronously
        /// </summary>
        /// <remarks>Returns immediately and will run the client-specified callback when the action
        /// later completes.  Any output arguments can then be retrieved by calling
        /// EndGetMessage().</remarks>
        /// <param name="aCallback">Delegate to run when the action completes.
        /// This is guaranteed to be run but may indicate an error</param>
        public void BeginGetMessage(CallbackAsyncComplete aCallback)
        {
            Invocation invocation = iService.Invocation(iActionGetMessage, aCallback);
            int outIndex = 0;
            invocation.AddOutput(new ArgumentString((ParameterString)iActionGetMessage.OutputParameter(outIndex++)));
            invocation.AddOutput(new ArgumentUint((ParameterUint)iActionGetMessage.OutputParameter(outIndex++)));
            iService.InvokeAction(invocation);
        }

        /// <summary>
        /// Retrieve the output arguments from an asynchronously invoked action.
        /// </summary>
        /// <remarks>This may only be called from the callback set in the above Begin function.</remarks>
        /// <param name="aAsyncHandle">Argument passed to the delegate set in the above Begin function</param>
        /// <param name="aMessage"></param>
        /// <param name="aMessageID"></param>
        public void EndGetMessage(IntPtr aAsyncHandle, out String aMessage, out uint aMessageID)
        {
            uint code;
            string desc;
            if (Invocation.Error(aAsyncHandle, out code, out desc))
            {
                throw new ProxyError(code, desc);
            }
            uint index = 0;
            aMessage = Invocation.OutputString(aAsyncHandle, index++);
            aMessageID = Invocation.OutputUint(aAsyncHandle, index++);
        }

        /// <summary>
        /// Set a delegate to be run when the Message state variable changes.
        /// </summary>
        /// <remarks>Callbacks may be run in different threads but callbacks for a
        /// CpProxyAvOpenhomeOrgMessageCenter1 instance will not overlap.</remarks>
        /// <param name="aMessageChanged">The delegate to run when the state variable changes</param>
        public void SetPropertyMessageChanged(System.Action aMessageChanged)
        {
            lock (iPropertyLock)
            {
                iMessageChanged = aMessageChanged;
            }
        }

        private void MessagePropertyChanged()
        {
            lock (iPropertyLock)
            {
                ReportEvent(iMessageChanged);
            }
        }

        /// <summary>
        /// Set a delegate to be run when the MessageID state variable changes.
        /// </summary>
        /// <remarks>Callbacks may be run in different threads but callbacks for a
        /// CpProxyAvOpenhomeOrgMessageCenter1 instance will not overlap.</remarks>
        /// <param name="aMessageIDChanged">The delegate to run when the state variable changes</param>
        public void SetPropertyMessageIDChanged(System.Action aMessageIDChanged)
        {
            lock (iPropertyLock)
            {
                iMessageIDChanged = aMessageIDChanged;
            }
        }

        private void MessageIDPropertyChanged()
        {
            lock (iPropertyLock)
            {
                ReportEvent(iMessageIDChanged);
            }
        }

        /// <summary>
        /// Query the value of the Message property.
        /// </summary>
        /// <remarks>This function is threadsafe and can only be called if Subscribe() has been
        /// called and a first eventing callback received more recently than any call
        /// to Unsubscribe().</remarks>
        /// <returns>Value of the Message property</returns>
        public String PropertyMessage()
        {
            PropertyReadLock();
            String val;
            try
            {
                val = iMessage.Value();
            }
            finally
            {
                PropertyReadUnlock();
            }
            return val;
        }

        /// <summary>
        /// Query the value of the MessageID property.
        /// </summary>
        /// <remarks>This function is threadsafe and can only be called if Subscribe() has been
        /// called and a first eventing callback received more recently than any call
        /// to Unsubscribe().</remarks>
        /// <returns>Value of the MessageID property</returns>
        public uint PropertyMessageID()
        {
            PropertyReadLock();
            uint val;
            try
            {
                val = iMessageID.Value();
            }
            finally
            {
                PropertyReadUnlock();
            }
            return val;
        }

        /// <summary>
        /// Must be called for each class instance.  Must be called before Core.Library.Close().
        /// </summary>
        public void Dispose()
        {
            lock (this)
            {
                if (iHandle == IntPtr.Zero)
                    return;
                DisposeProxy();
                iHandle = IntPtr.Zero;
            }
            iActionGetMessage.Dispose();
            iMessage.Dispose();
            iMessageID.Dispose();
        }
    }
}

File: OpenHome/Net/ControlPoint/Proxies/CpAvOpenhomeOrgMessageCenter1.cpp
#include "CpAvOpenhomeOrgMessageCenter1.h"
#include <OpenHome/Net/Core/CpProxy.h>
#include <OpenHome/Net/Private/CpiService.h>
#include <OpenHome/Private/Thread.h>
#include <OpenHome/Net/Private/AsyncPrivate.h>
#include <OpenHome/Net/Core/CpDevice.h>
#include <OpenHome/Net/Private/Error.h>
#include <OpenHome/Net/Private/CpiDevice.h>

namespace OpenHome {
namespace Net {

class SyncGetMessageAvOpenhomeOrgMessageCenter1 : public SyncProxyAction
{
public:
    SyncGetMessageAvOpenhomeOrgMessageCenter1(CpProxyAvOpenhomeOrgMessageCenter1& aProxy, Brh& aMessage, TUint& aMessageID);
    virtual void CompleteRequest(IAsync& aAsync);
private:
    CpProxyAvOpenhomeOrgMessageCenter1& iService;
    Brh& iMessage;
    TUint& iMessageID;
};

} // namespace Net
} // namespace OpenHome


using namespace OpenHome;
using namespace OpenHome::Net;


// SyncGetMessageAvOpenhomeOrgMessageCenter1

SyncGetMessageAvOpenhomeOrgMessageCenter1::SyncGetMessageAvOpenhomeOrgMessageCenter1(CpProxyAvOpenhomeOrgMessageCenter1& aProxy, Brh& aMessage, TUint& aMessageID)
    : iService(aProxy)
    , iMessage(aMessage)
    , iMessageID(aMessageID)
{
}

void SyncGetMessageAvOpenhomeOrgMessageCenter1::CompleteRequest(IAsync& aAsync)
{
    iService.EndGetMessage(aAsync, iMessage, iMessageID);
}


// CpProxyAvOpenhomeOrgMessageCenter1

CpProxyAvOpenhomeOrgMessageCenter1::CpProxyAvOpenhomeOrgMessageCenter1(CpDevice& aDevice)
    : iCpProxy("av-openhome-org", "MessageCenter", 1, aDevice.Device())
{
    OpenHome::Net::Parameter* param;

    iActionGetMessage = new Action("GetMessage");
    param = new OpenHome::Net::ParameterString("Message");
    iActionGetMessage->AddOutputParameter(param);
    param = new OpenHome::Net::ParameterUint("MessageID");
    iActionGetMessage->AddOutputParameter(param);

    Functor functor;
    functor = MakeFunctor(*this, &CpProxyAvOpenhomeOrgMessageCenter1::MessagePropertyChanged);
    iMessage = new PropertyString("Message", functor);
    AddProperty(iMessage);
    functor = MakeFunctor(*this, &CpProxyAvOpenhomeOrgMessageCenter1::MessageIDPropertyChanged);
    iMessageID = new PropertyUint("MessageID", functor);
    AddProperty(iMessageID);
}

CpProxyAvOpenhomeOrgMessageCenter1::~CpProxyAvOpenhomeOrgMessageCenter1()
{
    DestroyService();
    delete iActionGetMessage;
}

void CpProxyAvOpenhomeOrgMessageCenter1::SyncGetMessage(Brh& aMessage, TUint& aMessageID)
{
    SyncGetMessageAvOpenhomeOrgMessageCenter1 sync(*this, aMessage, aMessageID);
    BeginGetMessage(sync.Functor());
    sync.Wait();
}

void CpProxyAvOpenhomeOrgMessageCenter1::BeginGetMessage(FunctorAsync& aFunctor)
{
    Invocation* invocation = iCpProxy.GetService().Invocation(*iActionGetMessage, aFunctor);
    TUint outIndex = 0;
    const Action::VectorParameters& outParams = iActionGetMessage->OutputParameters();
    invocation->AddOutput(new ArgumentString(*outParams[outIndex++]));
    invocation->AddOutput(new ArgumentUint(*outParams[outIndex++]));
    iCpProxy.GetInvocable().InvokeAction(*invocation);
}

void CpProxyAvOpenhomeOrgMessageCenter1::EndGetMessage(IAsync& aAsync, Brh& aMessage, TUint& aMessageID)
{
    ASSERT(((Async&)aAsync).Type() == Async::eInvocation);
    Invocation& invocation = (Invocation&)aAsync;
    ASSERT(invocation.Action().Name() == Brn("GetMessage"));

    Error::ELevel level;
    TUint code;
    const TChar* ignore;
    if (invocation.Error(level, code, ignore)) {
        THROW_PROXYERROR(level, code);
    }
    TUint index = 0;
    ((ArgumentString*)invocation.OutputArguments()[index++])->TransferTo(aMessage);
    aMessageID = ((ArgumentUint*)invocation.OutputArguments()[index++])->Value();
}

void CpProxyAvOpenhomeOrgMessageCenter1::SetPropertyMessageChanged(Functor& aFunctor)
{
    iCpProxy.GetLock().Wait();
    iMessageChanged = aFunctor;
    iCpProxy.GetLock().Signal();
}

void CpProxyAvOpenhomeOrgMessageCenter1::SetPropertyMessageIDChanged(Functor& aFunctor)
{
    iCpProxy.GetLock().Wait();
    iMessageIDChanged = aFunctor;
    iCpProxy.GetLock().Signal();
}

void CpProxyAvOpenhomeOrgMessageCenter1::PropertyMessage(Brhz& aMessage) const
{
    AutoMutex a(iCpProxy.PropertyReadLock());
    if (iCpProxy.GetSubscriptionStatus() != CpProxy::eSubscribed) {
        THROW(ProxyNotSubscribed);
    }
    aMessage.Set(iMessage->Value());
}

void CpProxyAvOpenhomeOrgMessageCenter1::PropertyMessageID(TUint& aMessageID) const
{
    AutoMutex a(iCpProxy.PropertyReadLock());
    if (iCpProxy.GetSubscriptionStatus() != CpProxy::eSubscribed) {
        THROW(ProxyNotSubscribed);
    }
    aMessageID = iMessageID->Value();
}

void CpProxyAvOpenhomeOrgMessageCenter1::MessagePropertyChanged()
{
    ReportEvent(iMessageChanged);
}

void CpProxyAvOpenhomeOrgMessageCenter1::MessageIDPropertyChanged()
{
    ReportEvent(iMessageIDChanged);
}


void CpProxyAvOpenhomeOrgMessageCenter1::Subscribe()
{
  iCpProxy.Subscribe();
}

void CpProxyAvOpenhomeOrgMessageCenter1::Unsubscribe()
{
 iCpProxy.Unsubscribe();
}

void CpProxyAvOpenhomeOrgMessageCenter1::SetPropertyChanged(Functor& aFunctor)
{
  iCpProxy.SetPropertyChanged(aFunctor);
}

void CpProxyAvOpenhomeOrgMessageCenter1::SetPropertyInitialEvent(Functor& aFunctor)
{
  iCpProxy.SetPropertyInitialEvent(aFunctor);
}
void CpProxyAvOpenhomeOrgMessageCenter1::AddProperty(Property* aProperty)
{
  iCpProxy.AddProperty(aProperty);
}

void CpProxyAvOpenhomeOrgMessageCenter1::DestroyService()
{
  iCpProxy.DestroyService();
}

void CpProxyAvOpenhomeOrgMessageCenter1::ReportEvent(Functor aFunctor)
{
  iCpProxy.ReportEvent(aFunctor);
}

TUint CpProxyAvOpenhomeOrgMessageCenter1::Version() const
{
  return iCpProxy.Version();
}


